File: model_1/icpc-jakarta-2017/country_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <string>

using namespace std;

map<char, vector<int>> build_pos_map(const string& s) {
    map<char, vector<int>> mp;
    for (int i = 0; i < s.size(); i++) {
        mp[s[i]].push_back(i);
    }
    return mp;
}

string greedy_suffix(const string& S, int start, int length, const map<char, vector<int>>& pos_map, int n) {
    if (length == 0) {
        return "";
    }
    string res;
    int current = start;
    for (int j = 0; j < length; j++) {
        int R_bound = n - (length - j - 1);
        if (current > R_bound) {
            return "";
        }
        char found_char = 0;
        int found_index = n + 10000;
        for (char c = 'A'; c <= 'Z'; c++) {
            auto it = pos_map.find(c);
            if (it == pos_map.end()) {
                continue;
            }
            const vector<int>& arr = it->second;
            auto it2 = lower_bound(arr.begin(), arr.end(), current);
            if (it2 != arr.end() && *it2 <= R_bound) {
                found_char = c;
                found_index = *it2;
                break;
            }
        }
        if (found_char == 0) {
            return "";
        }
        res += found_char;
        current = found_index + 1;
    }
    return res;
}

string find_code_geq(const string& S, const string& prev, int K, const map<char, vector<int>>& pos_map) {
    int n = S.size();
    if (n < K) {
        return "";
    }
    if (S[0] < prev[0]) {
        return "";
    }
    string res = "";
    res += S[0];
    int current_index = 0;
    bool bound = (S[0] == prev[0]);

    for (int j = 1; j < K; j++) {
        int L = current_index + 1;
        int R_bound = n - (K - j);
        if (L > R_bound) {
            return "";
        }

        char low_bound = bound ? prev[j] : 'A';
        char found_char = 0;
        int found_index = n + 10000;
        for (char c = low_bound; c <= 'Z'; c++) {
            auto it = pos_map.find(c);
            if (it == pos_map.end()) {
                continue;
            }
            const vector<int>& arr = it->second;
            auto it2 = lower_bound(arr.begin(), arr.end(), L);
            if (it2 != arr.end() && *it2 <= R_bound) {
                found_char = c;
                found_index = *it2;
                break;
            }
        }
        if (found_char == 0) {
            return "";
        }
        res += found_char;
        current_index = found_index;
        if (bound) {
            if (found_char > prev[j]) {
                bound = false;
            }
        }
    }
    return res;
}

string next_code_helper(const string& S, const string& prev, int K, const map<char, vector<int>>& pos_map) {
    int n = S.size();
    vector<int> occ;
    int current_index = -1;
    for (int i = 0; i < K; i++) {
        char c = prev[i];
        auto it = pos_map.find(c);
        if (it == pos_map.end()) {
            return "";
        }
        const vector<int>& arr = it->second;
        auto it2 = lower_bound(arr.begin(), arr.end(), current_index + 1);
        if (it2 == arr.end()) {
            return "";
        }
        int pos_val = *it2;
        occ.push_back(pos_val);
        current_index = pos_val;
    }

    for (int i = K - 1; i >= 0; i--) {
        int L_bound;
        if (i == 0) {
            L_bound = 0;
        } else {
            L_bound = occ[i - 1] + 1;
        }
        int R_bound = n - (K - i - 1);
        if (L_bound > R_bound) {
            continue;
        }

        char current_char = prev[i];
        for (char c = current_char + 1; c <= 'Z'; c++) {
            auto it = pos_map.find(c);
            if (it == pos_map.end()) {
                continue;
            }
            const vector<int>& arr = it->second;
            auto it2 = lower_bound(arr.begin(), arr.end(), L_bound);
            if (it2 == arr.end() || *it2 > R_bound) {
                continue;
            }
            int pos_val = *it2;
            string suffix = greedy_suffix(S, pos_val + 1, K - i - 1, pos_map, n);
            if (suffix != "") {
                string new_code = prev.substr(0, i) + string(1, c) + suffix;
                return new_code;
            }
        }
    }
    return "";
}

int main() {
    int N, K_val;
    cin >> N >> K_val;
    vector<string> country_names;
    string line;
    getline(cin, line);
    for (int i = 0; i < N; i++) {
        getline(cin, line);
        country_names.push_back(line);
    }

    vector<map<char, vector<int>>> pos_maps;
    for (int i = 0; i < N; i++) {
        pos_maps.push_back(build_pos_map(country_names[i]));
    }

    vector<pair<string, int>> sorted_countries;
    for (int i = 0; i < N; i++) {
        sorted_countries.push_back({country_names[i], i});
    }
    sort(sorted_countries.begin(), sorted_countries.end());

    vector<string> ans(N, "");
    string prev_code = "";

    for (int idx = 0; idx < N; idx++) {
        string name = sorted_countries[idx].first;
        int orig_index = sorted_countries[idx].second;
        const map<char, vector<int>>& pm = pos_maps[orig_index];

        if (idx == 0) {
            string bound0 = string(1, name[0]);
            if (K_val > 1) {
                bound0 += string(K_val - 1, 'A');
            }
            string code = find_code_geq(name, bound0, K_val, pm);
            if (code == "") {
                cout << "NO" << endl;
                return 0;
            }
            ans[orig_index] = code;
            prev_code = code;
        } else {
            string candidate = find_code_geq(name, prev_code, K_val, pm);
            if (candidate == "") {
                cout << "NO" << endl;
                return 0;
            }
            if (candidate > prev_code) {
                ans[orig_index] = candidate;
                prev_code = candidate;
            } else {
                string next_candidate = next_code_helper(name, prev_code, K_val, pm);
                if (next_candidate == "") {
                    cout << "NO" << endl;
                    return 0;
                }
                ans[orig_index] = next_candidate;
                prev_code = next_candidate;
            }
        }
    }

    cout << "YES" << endl;
    for (int i = 0; i < N; i++) {
        cout << ans[i] << endl;
    }

    return 0;
}

File: model_1/icpc-jakarta-2017/country_1_solution.py
import bisect

def find_code_geq(S, t, K, pos_map):
    n = len(S)
    if S[0] < t[0]:
        return None
    res = [S[0]]
    current_index = 0
    bound = (S[0] == t[0])
    
    for j in range(1, K):
        L = current_index + 1
        R = n - (K - j)
        if L > R:
            return None
        
        if bound:
            low_bound = t[j]
        else:
            low_bound = 'A'
        
        candidate_char = None
        candidate_index = None
        for c_code in range(ord(low_bound), ord('Z') + 1):
            c_char = chr(c_code)
            if c_char not in pos_map:
                continue
            arr = pos_map[c_char]
            pos_in_arr = bisect.bisect_left(arr, L)
            if pos_in_arr < len(arr) and arr[pos_in_arr] <= R:
                candidate_char = c_char
                candidate_index = arr[pos_in_arr]
                break
        
        if candidate_char is None:
            return None
        
        res.append(candidate_char)
        current_index = candidate_index
        
        if bound:
            if candidate_char > t[j]:
                bound = False
                
    return ''.join(res)

def greedy_suffix(S, start, length, pos_map, n):
    if length == 0:
        return ""
    res = []
    current = start
    for j in range(length):
        R_bound = n - (length - j)
        found_char = None
        found_index = None
        for c_code in range(ord('A'), ord('Z') + 1):
            c_char = chr(c_code)
            if c_char not in pos_map:
                continue
            arr = pos_map[c_char]
            pos_in_arr = bisect.bisect_left(arr, current)
            if pos_in_arr < len(arr) and arr[pos_in_arr] <= R_bound:
                found_char = c_char
                found_index = arr[pos_in_arr]
                break
        if found_char is None:
            return None
        res.append(found_char)
        current = found_index + 1
    return ''.join(res)

def next_code_helper(S, t, K, pos_map):
    n = len(S)
    occ = []
    current_index = -1
    for i in range(K):
        c = t[i]
        if c not in pos_map:
            return None
        arr = pos_map[c]
        pos_in_arr = bisect.bisect_right(arr, current_index)
        if pos_in_arr >= len(arr):
            return None
        occ_val = arr[pos_in_arr]
        occ.append(occ_val)
        current_index = occ_val

    for i in range(K - 1, 0, -1):
        L = occ[i - 1] + 1
        R = n - (K - i)
        if L > R:
            continue
        candidate_char = None
        candidate_index = None
        for c_code in range(ord(t[i]) + 1, ord('Z') + 1):
            c_char = chr(c_code)
            if c_char not in pos_map:
                continue
            arr = pos_map[c_char]
            pos_in_arr = bisect.bisect_left(arr, L)
            if pos_in_arr < len(arr) and arr[pos_in_arr] <= R:
                candidate_char = c_char
                candidate_index = arr[pos_in_arr]
                break
        if candidate_char is None:
            continue
        suffix = greedy_suffix(S, candidate_index + 1, K - i - 1, pos_map, n)
        if suffix is None:
            continue
        new_code = t[:i] + candidate_char + suffix
        return new_code
    return None

def main():
    import sys
    data = sys.stdin.read().splitlines()
    if not data: 
        return
    n_countries, K_val = map(int, data[0].split())
    country_names = []
    for i in range(1, 1 + n_countries):
        country_names.append(data[i].strip())
    
    pos_maps = {}
    for name in country_names:
        d = {}
        for idx, char in enumerate(name):
            if char not in d:
                d[char] = []
            d[char].append(idx)
        pos_maps[name] = d
    
    sorted_countries = sorted([(name, idx) for idx, name in enumerate(country_names)], key=lambda x: x[0])
    
    codes = [None] * n_countries
    prev_code = "A" * K_val
    
    for i in range(n_countries):
        name, orig_idx = sorted_countries[i]
        if i == 0:
            candidate = find_code_geq(name, prev_code, K_val, pos_maps[name])
            if candidate is None:
                print("NO")
                return
            codes[orig_idx] = candidate
            prev_code = candidate
        else:
            candidate = find_code_geq(name, prev_code, K_val, pos_maps[name])
            if candidate is None:
                print("NO")
                return
            if candidate > prev_code:
                codes[orig_idx] = candidate
                prev_code = candidate
            else:
                next_candidate = next_code_helper(name, prev_code, K_val, pos_maps[name])
                if next_candidate is None:
                    print("NO")
                    return
                codes[orig_idx] = next_candidate
                prev_code = next_candidate
                
    print("YES")
    for code in codes:
        print(code)

if __name__ == "__main__":
    main()